### ISSUE TITLE
[Profiler] Issues in alignment for ExecutionTrace: rf_id and Kineto: external id.

### ISSUE THREAD

**Issue by JoongunPark** [pytorch/pytorch#122833]
### 🐛 Describe the bug

This Pull Request ([#119912](https://github.com/pytorch/pytorch/pull/119912)) introduces functionality to simultaneously gather Execution Trace and Kineto trace data. It provides on correlating the `rf_id `from Execution Trace with the `external id` from Kineto trace, a step critical for the effective analysis of the collected traces.

Key issues found from this PR are:

1) A consistent offset observed between `rf_id `and `external_id`.
2) Misalignment of rf_id and external_id identifiers occurring after the `[pytorch|profiler|execution_trace|thread]` annotation.

For demonstration and testing purposes, here is a Resnet-50 example available at [resnet50.py](https://github.com/JoongunPark/kineto/blob/main/tb_plugin/examples/new_resnet50.py). 
To execute the example and collect both types of traces, run the following command:
```
# Download 
$ git clone https://github.com/JoongunPark/kineto
$ pip3 install -r requirements.txt

# Install pytorch nightly version (check your cuda version / need to use python>=3.10)
$ pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu118
or
$ pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu121

$ cd jupark_kineto/tb_plugin/examples/
$ mkdir ./result/

# Update number of GPUs
$ vi new_resnet50.py

119 if __name__ == "__main__":
120     size = 2 # size should be updated as number of GPUs you have. 
121     processes = []
122     mp.set_start_method("spawn")
123     for rank in range(size):
124         p = mp.Process(target=init_process, args=(rank, size, example))
125         p.start()
126         processes.append(p)
127
128     for p in processes:
129         p.join()
 

# Run
$ python3 new_resnet50.py

# Result
$ cd ./result/
$ ls
```

### Collected Trace
[Resnet-50_2GPU_GTX1070.zip](https://github.com/pytorch/pytorch/files/14782011/Resnet-50_2GPU_GTX1070.zip)

Misalignment can be found after PyTorch op ID: 3710
```
[2024-03-27 18:55:29,873] trace_link.py:802 [WARNING]: No corresponding Kineto op found for PyTorch op ID: 3710, Name: '[pytorch|profiler|execution_trace|thread]'.
[2024-03-27 18:55:29,874] trace_link.py:802 [WARNING]: No corresponding Kineto op found for PyTorch op ID: 3711, Name: 'autograd::engine::evaluate_function: NllLossBackward0'.
[2024-03-27 18:55:29,875] trace_link.py:802 [WARNING]: No corresponding Kineto op found for PyTorch op ID: 3712, Name: 'NllLossBackward0'.
```

### Versions

Collecting environment information...
PyTorch version: 2.4.0.dev20240326+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.29.0
Libc version: glibc-2.31

Python version: 3.10.13 (main, Aug 25 2023, 13:20:03) [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-97-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce GTX 1070
GPU 1: NVIDIA GeForce GTX 1070

Nvidia driver version: 550.54.14
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      46 bits physical, 48 bits virtual
CPU(s):                             12
On-line CPU(s) list:                0-11
Thread(s) per core:                 2
Core(s) per socket:                 6
Socket(s):                          1
NUMA node(s):                       1
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              63
Model name:                         Intel(R) Core(TM) i7-5930K CPU @ 3.50GHz
Stepping:                           2
CPU MHz:                            3599.881
CPU max MHz:                        3700.0000
CPU min MHz:                        1200.0000
BogoMIPS:                           6999.77
L1d cache:                          192 KiB
L1i cache:                          192 KiB
L2 cache:                           1.5 MiB
L3 cache:                           15 MiB
NUMA node0 CPU(s):                  0-11
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb invpcid_single pti intel_ppin ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_llc dtherm ida arat pln pts md_clear flush_l1d

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] pytorch-triton==3.0.0+989adb9a29
[pip3] torch==2.4.0.dev20240326+cu118
[pip3] torchaudio==2.2.0.dev20240327+cu118
[pip3] torchvision==0.19.0.dev20240327+cu118
[conda] blas                      1.0                         mkl
[conda] mkl                       2021.2.0           h06a4308_296
[conda] mkl-service               2.3.0            py38h27cfd23_1
[conda] mkl_fft                   1.3.0            py38h42c9631_2
[conda] mkl_random                1.2.1            py38ha9443f7_2
[conda] numpy                     1.20.1           py38h93e21f0_0
[conda] numpy-base                1.20.1           py38h7d8b39e_0
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1

cc @robieta @chaekit @aaronenyeshi @nbcsm @guotuofeng @guyang3532 @gaoteng-git @tiffzhaofb @dzhulgakov @davidberard98

**Comment by briancoutinho** [comment]
@JoongunPark  thanks for filing the issue :) 
TLDR - I actually dug into the code for the pytorch profiler and discovered that "external id" is in fact not related to "record function iD". In fact that path won't work. But, I realized its quite easy (2-3 line change) to add "record function id " itself as an attribute in the profiler trace. So that works!

### Overview - record function ID
* Record function is the mechanism in PyTorch to subscribe to function start and exit. Both ET and Kineto/Profiler are using it. The execution trace is picking up an ID called ["handle()"](https://github.com/pytorch/pytorch/blob/ed37fbdf6030e940c046207b7057d3962743d731/aten/src/ATen/record_function.h#L410) from record function that is incremented [here](https://github.com/pytorch/pytorch/blob/ed37fbdf6030e940c046207b7057d3962743d731/aten/src/ATen/record_function.cpp#L536).
* This handle is [logged by ET](https://github.com/pytorch/pytorch/blob/ed37fbdf6030e940c046207b7057d3962743d731/torch/csrc/profiler/standalone/execution_trace_observer.cpp#L605) inside the observer.

### What is external ID
Ok so now let's dig into how "External ID" appears in Kineto traces. I read up from Kineto first and what I discovered is this is "correlationID" for pytorch operators. This is in kineto repo, let me know if you want pointers.

We can look up where the PyTorch operator events are transferred from PyTorch to Kineto. This part of the code turns out to be quite complex and it took me a while to debug and understand it.
(You can skip this, its more informative)
The two files of interest are profiler/collection.cpp and autograd/profiler_kineto.cpp. Profiler has optimized queues per thread that store TorchOp - operator event information. At the end it finalizes and move these data into Kineto. We can work backwards from kineto-
1. The functions `passEventsToKineto()` is basically adding PyTorch operator events to a buffer and transferring it to kineto [here](https://github.com/pytorch/pytorch/blob/ed37fbdf6030e940c046207b7057d3962743d731/torch/csrc/profiler/collection.cpp#L803-L821). If we look closely there is a "e->correlationID()", this "e" is coming from a data structure called `Result`.  Each operation/kernel etc are saved as a Result. The design of Result is a bit complex, it has a templated `extra_fields_` object member that actually contains additional information. This member changes based on the operation type like if it is a TorchOp. [Result defn](https://github.com/pytorch/pytorch/blob/ed37fbdf6030e940c046207b7057d3962743d731/torch/csrc/profiler/collection.h#L400).
2.  Although a bit convoluted, the Result's correlation ID comes from "TorchOps" correlation ID, shown [here](https://github.com/pytorch/pytorch/blob/ed37fbdf6030e940c046207b7057d3962743d731/torch/csrc/profiler/collection.h#L400).  
3. We can find where this data member is [assigned here](https://github.com/pytorch/pytorch/blob/ed37fbdf6030e940c046207b7057d3962743d731/torch/csrc/profiler/collection.cpp#L447). This gets it from the funky function `ThreadLocalSubqueue::TorchOpStorage::OpList::correlationID`. 
4. After doing some vodoo magic here its actually coming from this EventBlock::correlation_id() function

https://github.com/pytorch/pytorch/blob/ed37fbdf6030e940c046207b7057d3962743d731/torch/csrc/profiler/collection.cpp#L295C1-L301C2
```
template <typename T, size_t ChunkSize>
uint64_t ThreadLocalSubqueue::TorchOpStorage::EventBlock<T, ChunkSize>::
    correlation_id(const T* ptr) const {
  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
      ptr >= this->data() && ptr < this->data() + ChunkSize);
  return id_start_ + (ptr - this->data());
}
```
As we can see in the code this is some internal unique ID and probably not going to always match rf_id.

### Directly logging record Function ID

Ok, this brings me to my final change, we can actually just log the record function ID itself. I did a quick hack change and i am able to log it in profiler direclty. So we can just read this "Record function ID" field instead of "External ID"
https://github.com/pytorch/pytorch/pull/122948

**Comment by briancoutinho** [comment]
Here are some early results using https://github.com/pytorch/pytorch/pull/122948 
  `pytest test/profiler/test_profiler.py -k test_execution_trace_with_kineto`

 I can see the new record function ID  field in ET and Kineto
 
![Screenshot 2024-03-28 at 5 50 28 PM](https://github.com/pytorch/pytorch/assets/6922212/459b5108-fa85-4b75-9cc6-deb505834d61)
![Screenshot 2024-03-28 at 5 50 34 PM](https://github.com/pytorch/pytorch/assets/6922212/9f4e8d0c-afda-4ca8-9642-afef0319c3e0)

These match even if the "external ID" does not
![Screenshot 2024-03-28 at 5 48 55 PM](https://github.com/pytorch/pytorch/assets/6922212/f7c13aa2-4fd6-4b29-88f7-6cc559fc46c7)
![Screenshot 2024-03-28 at 5 49 14 PM](https://github.com/pytorch/pytorch/assets/6922212/19b96f24-9572-47c3-8c0b-b09dcd65c1dd)
  

To build I suggest checking out pytorch, then checkout my branch "git checkout profiler_add_rf_id"
Follow build instructions https://github.com/pytorch/pytorch/tree/main?tab=readme-ov-file#get-the-pytorch-source
I used this command
`DEBUG=1 USE_DISTRIBUTED=0 USE_MKLDNN=0 USE_CUDA=1 BUILD_TEST=0 USE_FBGEMM=0 USE_NNPACK=0 USE_QNNPACK=0 USE_XNNPACK=0 python setup.py develop`
